kafka-client/src/main/java/dev/responsive/kafka/api/async/internals/metrics/AsyncProcessorMetricsRecorder.java

```diff
@@ -0,0 +1,146 @@
+  public static final String SCHEDULING_QUEUE_SIZE_DESC = "entries in the scheduling queue";
+  public static final String SCHEDULING_QUEUE_LONGEST_SIZE = "scheduling-queue-longest-size";
+  public static final String SCHEDULING_QUEUE_LONGEST_SIZE_DESC
+      = "entries in the queue for key w/ most entries in scheduling queue";
+
+  public static final String FROM_STATE = "from-state";
+  public static final String TO_STATE = "to-state";
+
+  private final ResponsiveMetrics.MetricScope scope;
+  private final ResponsiveMetrics metrics;
+  private final MetricName pendingEventsGauge;
+  private final Sensor eventProcessSensor;
+  private final Sensor schedulingQueueSizeSensor;
+  private final Sensor schedulingQueueLongestQueueSizeSensor;
+  private final ConcurrentMap<String, Sensor> stateTransitionSensors = new ConcurrentHashMap<>();
+
+  public AsyncProcessorMetricsRecorder(
+      final String threadId,
+      final TaskId taskId,
+      final String processorName,
+      final ResponsiveMetrics metrics,
+      final Supplier<Integer> pendingEventsSize
+  ) {
+    this.metrics = metrics;
+    scope = metrics.processorLevelMetric(
+        GROUP_NAME, threadId, taskId, processorName);
+
+    this.pendingEventsGauge = metrics.metricName(PENDING_EVENTS, PENDING_EVENTS_DESC, scope);
+    metrics.addMetric(pendingEventsGauge, (Gauge<Integer>) (c, n) -> pendingEventsSize.get());
+
+    this.eventProcessSensor = metrics.addSensor(scope.sensorName(PROCESS_TIME));
+    eventProcessSensor.add(
+        metrics.metricName(PROCESS_TIME_AVG, PROCESS_TIME_AVG_DESC, scope),
+        new Avg()
+    );
+    eventProcessSensor.add(
+        metrics.metricName(PROCESS_TIME_MAX, PROCESS_TIME_MAX_DESC, scope),
+        new Max()
+    );
+
+    schedulingQueueSizeSensor = metrics.addSensor(scope.sensorName(SCHEDULING_QUEUE_SIZE));
+    schedulingQueueSizeSensor.add(
+        metrics.metricName(SCHEDULING_QUEUE_SIZE, SCHEDULING_QUEUE_SIZE_DESC, scope),
+        new Value()
+    );
+
+    schedulingQueueLongestQueueSizeSensor
+        = metrics.addSensor(scope.sensorName(SCHEDULING_QUEUE_LONGEST_SIZE));
+    schedulingQueueLongestQueueSizeSensor.add(
+        metrics.metricName(
+            SCHEDULING_QUEUE_LONGEST_SIZE,
+            SCHEDULING_QUEUE_LONGEST_SIZE_DESC,
+            scope
+        ),
+        new Value()
+    );
+  }
+
+  public void recordEventProcess(final Duration duration) {
+    eventProcessSensor.record(duration.toNanos());
+  }
+
+  public void recordSchedulingQueueSize(final int size) {
+    schedulingQueueSizeSensor.record(size);
+  }
+
+  public void recordSchedulingQueueLongestSize(final int size) {
+    schedulingQueueLongestQueueSizeSensor.record(size);
+  }
+
+  public void recordStateTransition(
+      final AsyncEvent.State from,
+      final Instant fromTime,
+      final AsyncEvent.State to,
+      final Instant toTime
+  ) {
+    final var innerScope = scope
```

Comment:
This is very nitty so feel free to ignore, but imo some things are bit difficult to follow when everything's declared as a var. Like this line. Probably my own fault for reviewing directly on github instead of pulling in the PR locally but sometimes I'm lazy 🙂

Reply:
nah that's fair. Putting a type here doesn't hurt.